<a href="https://colab.research.google.com/github/NicolasZarateC/Analitica-de-Negocios/blob/main/Reto_1_Car_Crash.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

**Reto 1: Car Crash** - Nicolas Zarate

In [ ]:
import numpy as np
import pandas as pd

In [ ]:
nxl= "/content/sample_data/1. BD2_CarCrash_Int.xlsx"
XDB=pd.read_excel(nxl)

XDB=XDB[['County','Weekday','ViolCat','ClearWeather','Month','CrashType','Highway','Daylight',]]

#Filtrar para que sea solo de SAN FRANCISCO
filtered_data = XDB[XDB['County'] == 'SAN FRANCISCO']
XDB = filtered_data
XDB['CrashType']=pd.factorize(XDB['CrashType'])[0] #Para pasar las letras a numeros de CrashType


In [ ]:
#Organizar Datos
XD=XDB[['Weekday','ViolCat','ClearWeather','Month','CrashType','Daylight',]]
yd=XDB[['Highway']]


In [ ]:
from os import ctermid
from sklearn.naive_bayes import GaussianNB

**Preguntas:**

In [ ]:
#¿Cuál es el número de datos total de la base de datos?
mnb=GaussianNB()
mnb.fit(XD,yd)

ct=mnb.class_count_
print("El total de los datos es:",len(XD))

#¿Cuál es el porcentaje de los datos que hacen parte de cada una de las categorías de clasificación?
print("Como estan separados los datos",ct)
print("Los datos porcentualmente estan divididos:", ct/len(XD))

u=mnb.theta_
print("La media de las variables de entrada por categoria es:\n",u)
std=mnb.var_
std=np.sqrt(std)
print("La desviacion estandar de las variables de entrada por categoria es:\n",std)

#Se identifican los valores superiores e inferiores
print("El limite superior de los valores:\n", u+std)
print("El limite inferior de los valores:\n", u-std)

#Determinar los intervalos en los cual se encuentra la variación del día de la semana para cada una de las categorías de clasificación
#El intervalo cuando el accidente no es en carretera entre 5.835 y 1.934
#El intervalo cuando el accidente es en carretera entre 5.942 y 1.910


#Determinar los intervalos en los cual se encuentra la variación de la ViolCat para cada una de las categorías de clasificación
#El intervalo cuando el accidente no es en carretera entre 10.978 y 3.936
#El intervalo cuando el accidente es en carretera entre 6.312 y 1.988


#Determinar el intervalo en los cuales se encuentra la variación de la Categoría de accidentes para cada una de las categorías de clasificación
#El intervalo cuando el accidente no es en carretera entre 5.091 y 1.344
#El intervalo cuando el accidente es en carretera entre 3.336 y 1.220


El total de los datos es: 2081
Como estan separados los datos [1580.  501.]
Los datos porcentualmente estan divididos: [0.75925036 0.24074964]
La media de las variables de entrada por categoria es:
 [[3.88417722 7.45696203 0.86265823 4.35189873 3.21772152 0.66265823]
 [3.9261477  4.1497006  0.79840319 6.69461078 2.27744511 0.62075848]]
La desviacion estandar de las variables de entrada por categoria es:
 [[1.95062542 3.52136997 0.34420782 2.3433645  1.87371962 0.47280262]
 [2.01603723 2.16198902 0.40119266 3.44163641 1.05765448 0.48519831]]
El limite superior de los valores:
 [[ 5.83480263 10.978332    1.20686605  6.69526323  5.09144114  1.13546084]
 [ 5.94218494  6.31168962  1.19959585 10.13624719  3.33509959  1.10595679]]
El limite inferior de los valores:
 [[1.9335518  3.93559205 0.51845041 2.00853424 1.3440019  0.18985561]
 [1.91011047 1.98771158 0.39721054 3.25297437 1.21979063 0.13556017]]


/usr/local/lib/python3.11/dist-packages/sklearn/utils/validation.py:1408: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, warn=True)


In [ ]:
ydp=mnb.predict(XD)

from sklearn.metrics import confusion_matrix
cm=confusion_matrix(yd,ydp)
print(cm)

VN=cm[0,0];FP=cm[0,1];FN=cm[1,0];VP=cm[1,1]

#¿Cuál es la exactitud alcanzada por el modelo?

Ex=(VN+VP)/(VN+VP+FN+FP)
print("La exactitud es:", Ex)

#¿Cuál es la Especificidad alcanzada por el modelo? - Indicar el intervalo en donde se ubica la cantidad de verdaderos negativo - Indicar el intervalo en donde se ubica la cantidad de verdaderos positivos
Esp=VN/(VN+FP)
print("La especificidad es:", Esp)
#VN es 1365 y #VP es 277

#¿Cuál es la tasa de error alcanzada por el modelo?
TE=(FN+FP)/(VN+VP+FN+FP)
print("La tasa de error es:", TE)


#¿Cuál es la Sensibilidad alcanzada por el modelo?
Sen=VP/(VP+FN)
print("La sensibilidad es:", Sen)


#¿Cuál es la Precisión Alcanzada por el Modelo?

pre=VP/(VP+FP)
print("La precision es:", pre)

#¿Indicar la categoría a la que pertenece una persona con las siguientes características [5,1,1,3,4,0]?
XS=[[5,1,1,3,4,0]]
mnb.predict(XS)

print("Un individuo con estas caracterisitcas tiene el siguiente comportamiento:", mnb.predict(XS))

[[1365  215]
 [ 224  277]]
La exactitud es: 0.7890437289764536
La especificidad es: 0.8639240506329114
La tasa de error es: 0.21095627102354636
La sensibilidad es: 0.5528942115768463
La precision es: 0.5630081300813008
Un individuo con estas caracterisitcas tiene el siguiente comportamiento: [0]


/usr/local/lib/python3.11/dist-packages/sklearn/utils/validation.py:2739: UserWarning: X does not have valid feature names, but GaussianNB was fitted with feature names
  warnings.warn(
/usr/local/lib/python3.11/dist-packages/sklearn/utils/validation.py:2739: UserWarning: X does not have valid feature names, but GaussianNB was fitted with feature names
  warnings.warn(


**Analysis**

De acuerdo con los resultados que nos da el modelo, podemos corraborar en la plataforma colab respecto al comportamiento del modelo Naive Bayes y el excel de refencia conteniendo información sobre los choques en carretera dentro de la ciudad de San Francisco. El modelo alcanzó una exactitud cercana al 79% y destaca con un promedio de especificidad del 86% y una tasa de error del 21% respecto a las otras caracteristicas que encontramos cuando un accidente pasa en carrertera o no.

Con un total de 2081 choques en San Francisco, podemos ver que un 79% de los accidentes en esta ciudad no fueron en carretera y el 24% si lo fueron, esto corresponde a un total de 1580 choques por fuera y 501 dentro de una carretera probandonos que hay más choques por fuera de carretera en esta ciudad que dentro de una.

Respecto a la variable días de semana y su relación con los accidentes en carretera, vemos que el intervalo cuando el accidente no es en carretera es entre 5.835 y 1.934, por otro lado el intervalo cuando el accidente es en carretera es entre 5.942 y 1.910 que corresponde a que una mayoria de accidentes pasan de martes a sabado tanto cuando pasan en una carretera o no. Cuando miramos los intervalos del nivel de impacto, vemos que cuando el accidente no es en carretera se encuentra entre 10.978 y 3.936 mientras que el de un accidente es en carretera es entre 6.312 y 1.988. Por último, mirando los limites de categoria de accidente vemos que el intervalo se encuentra entre  5.091 y 1.344  para los accidentes que no pasan en carretera y entre 3.336 y 1.220 mostrandonos que los accidentes no en carretera son golpes laterales, de extremo trasero, de costado, de golpe a objeto, o de volcadura, cuando los que son en carretera son estos mismos sin incluir el de volcadura.

Con una exactitud de 79%, y una especificidad del 86% podemos ver que el modelo nos brinda una información muy exacta respecto a los datos evaluados. Además, con una tasa de error del 21%, corraboramos que se puede confiar en este modelo debido a que la predicción de los errores es baja. También vemos que el modelo nos da una sensibilidad del 55% lo cual dice que el modelo puede identificar moderadamente los accidentes que occurieron en carretera en casos positivos. Finalmente con una precision del 56%, el modelo acierta esta porcentaje de veces las predicciones positivas del modelo.

Mirando mas la matriz de confusión, con un total de 1365 valores negativos vemos que el modelo clasificó esta cantidad de accidentes no ocurridos en carretera y con un 277 valores positivos vemos que el modelo nos pone correctamente accidentes ocurridos en carretera.

Una persona que se chocó en San Francisco un viernes (weekday 5) mientras conducia o montaba en bicicleta bajo la influencia de alcohol odrogas (ViolCat 1), en clima despejado (ClearWheather 1), en el mes de marzo (Month 3) de costado (CrashType 4) sin luz del dia (Daylight 0) no tiende a tener un choque en carretera.